### ISSUE TITLE
LUA usage documentation

### ISSUE THREAD

**Issue by adelosa** [HackerSmacker/Lua370#1]
Hi there,
Thanks for doing this.. I'm interested in MVS (I work with Z/OS systems) and C programming (have always want to learn).. I have already learned some things about using JCC from your work..

So I cloned this repo and managed to compile it using the JCC compiler. Was not too tricky -- a couple of makefile changes for my environment, thats it. It assumes lots of knowledge.. you need the JCC compiler.. where do I get that.. and what do I need to do to install it (add it to the path).. I know that because I have been trawling the sites for a while, but its not easy to find these details.
Also not sure what the next step is once you have the object files.. How do I upload the file to my TK4 machine... What does the JCL for the link edit look like... 
Then how can I use it.. Batch.. TSO.. what other thinks do I need to do to make this work..

I think this is just doco that needs to be in the repo.. these are the types of questions that an new user will be wanting to know.. Happy to help with that.. will submit some pull requests with this information once I figure it out myself.

My second thought is.. what is the approach for incorporating upstream changes from the LUA team? Right now I'm not 100 certain how that would work. I can see some obvious mass updates like commenting out <stddef.h> and adding <stdlib.h> but I can see there are other changes to cater for... maybe unsupported language elements in JCC?.. not sure.. 
I think this project should have a plan on how to take the upstream version of LUA and then compile it for MVS.. Of course there will be some patching required but you don't want to have to do a wholesale review everytime a new release comes out. And have the same source managed in 2 places...

Is the eventual target to have this included with the next TK4 distribution.. Makes the most sense for me.. not sure what your thoughts are there.

Let me know your thoughts on the above.

**Comment by HackerSmacker** [comment]
Hi, thanks for the feedback. 
1. I definitely need to add better build directions for sure.
2. Lua/370 runs on TSO, and in batch!
3. I don't plan on actually trying to submit my changes to Lua/370 upstream, as this is a MVS-centric version that will have a lot of specific features that only work on this platform.
4. Ideally yes, but there are tons of changes on Lua/370 that make it unfit for a "rolling release", only for one-off forks and compilations. 
5. Eventually yes, I hope to add Lua/370 to TK4- when it is mature enough to be added.

**Comment by adelosa** [comment]
On item 3, I was not suggesting that the changes go back upstream. I was
wondering what the strategy is to release new versions of Lua/370 when
there is a new Lua release/update upstream. Having some approach that does
not require diffing the whole base looking for changes would be good.

Also what version of Lua is the base?

**Comment by adelosa** [comment]
Hi again.. 
Just trying to get a basic LUA batch job running, hoping you tell me if I am heading in the right direction...
I have compiled the source, pushed lua_linked.obj and luac_linked.obj up to the mainframe (in binary)
I then linked these modules.. using IEWL.. similar for LUA

    //LINK1    EXEC PGM=IEWL,PARM='LET,LIST,XREF,MAP'
    //SYSUT1   DD UNIT=VIO,SPACE=(CYL,(1,1))
    //SYSPRINT DD SYSOUT=*
    //SYSLMOD  DD DSN=HERC01.TEST.LOADLIB(LUAC),DISP=SHR
    //SYSLIN   DD DSN=HERC01.TEST.ASM(LUACOBJ),DISP=SHR
    /*

So I now have the loadlibs in HERC01.TEST.LOADLIB.. Both LUA and LUAC

I then submit the following JCL to run a LUA hello world
```
//HERC01LR JOB  (SETUP),                        
//             'RUN LUA MODULE',                
//             CLASS=A,                         
//             MSGCLASS=H,                      
//             MSGLEVEL=(1,1),                  
//             NOTIFY=HERC01                    
//LUARUN   EXEC PGM=LUA                         
//SYSPRINT DD SYSOUT=*                          
//SYSTERM  DD SYSOUT=*                          
//STEPLIB  DD DSN=HERC01.TEST.LOADLIB,DISP=SHR  
//SYSIN    DD *                                 
print("Hello world")                            
/*                                              
```
The job completes with 0 return code but no output in SYSPRINT or SYSTERM.. am I supposed to do something different or should this work?

**Comment by HackerSmacker** [comment]
So, after linking with IEWL, you're good to go. Also, the DD names are slightly different: STDOUT, STDIN, and STDERR. Define those, and you should see some output.

**Comment by adelosa** [comment]
Hi again.. ok.. so I am now getting some output.. but still not quite there..

My amended JCL.. Had to add a region statement as the first run blew the memory.

```
//LUARUN   EXEC PGM=LUA,REGION=1024K          
//STEPLIB  DD DSN=HERC01.TEST.LOADLIB,DISP=SHR
//STDOUT   DD SYSOUT=*                        
//STDERR   DD SYSOUT=*                        
//STDIN    DD *                               
print("Hello world")                          
/*                                            
```
Output in STDOUT

```
Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio 
> >                                                 
stdin:1: unexpected symbol near '00001202'          
```
So that seems like it does not like my program.. 
Tried adding some blank lines either side of the sysin.. that caused the CPU usage to go crazy (some kind of loop).. 
So the question is, have I managed to compile a dodgy binary or am I still doing something wrong..

All this will be included in a new readme for the project I am currently creating.

**Comment by adelosa** [comment]
Thought I would add a PARM to make the program dd explicit -
Adding a PARM='dd:sysin' to execute the LUA script at //SYSIN DD results in the following error..

    LUA: cannot open dd:sysin: Permission denied

I'm using a stock TK4 distribution using user HERC01 using datasets owned by HERC01... so don't feel like that is right.
Tried an inline file as well as a PDS member.. just does not want to access the file..

Running with PARM='-v' provides the version information so pretty OK that the binary is OK.

    Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio

**Comment by HackerSmacker** [comment]
Yeah, I'm still trying to figure that one out. In theory, you should pass the "--" argument and put your program on STDIN. Then, it should work.
